### PR TITLE
feat(ingest): implement conditional updates for rest endpoints

### PR DESCRIPTION
**Pull request by djordje-mijatovic**
## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: li-utils/src/main/java/com/linkedin/metadata/Constants.java [hunk 1]
```diff
@@ -15,6 +15,8 @@ public class Constants {
 
   public static final String DEFAULT_RUN_ID = "no-run-id-provided";
 
+  public static final String IN_UNMODIFIED_SINCE = "If-Unmodified-Since";
```
Comment (mattmatravers) [inline li-utils/src/main/java/com/linkedin/metadata/Constants.java:18]:
```suggestion
  public static final String IF_UNMODIFIED_SINCE = "If-Unmodified-Since";
```

File: metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/entities/EntitiesController.java [hunk 2]
```diff
@@ -103,15 +107,17 @@ public ResponseEntity<UrnResponseMap> getEntities(
 
   @PostMapping(value = "/", produces = MediaType.APPLICATION_JSON_VALUE)
   public ResponseEntity<List<String>> postEntities(
-      @RequestBody @Nonnull List<UpsertAspectRequest> aspectRequests) {
+      @RequestBody @Nonnull List<UpsertAspectRequest> aspectRequests,
+      @RequestHeader(name = Constants.IN_UNMODIFIED_SINCE) String condUpdate) {
```
Comment (mattmatravers) [inline metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/entities/EntitiesController.java:111]:
```suggestion
      @RequestHeader(name = Constants.IF_UNMODIFIED_SINCE) String condUpdate) {
```

File: metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/platform/entities/PlatformEntitiesController.java [hunk 3]
```diff
@@ -43,14 +47,16 @@ public void initBinder(WebDataBinder binder) {
 
   @PostMapping(value = "/", produces = MediaType.APPLICATION_JSON_VALUE)
   public ResponseEntity<List<String>> postEntities(
-      @RequestBody @Nonnull List<MetadataChangeProposal> metadataChangeProposals) {
+      @RequestBody @Nonnull List<MetadataChangeProposal> metadataChangeProposals,
+      @RequestHeader(name = Constants.IN_UNMODIFIED_SINCE) String condUpdate) {
```
Comment (mattmatravers) [inline metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/platform/entities/PlatformEntitiesController.java:51]:
```suggestion
      @RequestHeader(name = Constants.IF_UNMODIFIED_SINCE) String condUpdate) {
```

File: metadata-io/src/main/java/com/linkedin/metadata/entity/cassandra/CassandraAspectDao.java [hunk 4]
```diff
@@ -545,8 +546,10 @@ private SimpleStatement generateSaveStatement(EntityAspect aspect, boolean inser
               .value(CassandraAspect.CREATED_ON_COLUMN, literal(aspect.getCreatedOn().getTime()))
               .value(CassandraAspect.CREATED_FOR_COLUMN, literal(aspect.getCreatedFor()))
               .value(CassandraAspect.ENTITY_COLUMN, literal(entity))
-              .value(CassandraAspect.CREATED_BY_COLUMN, literal(aspect.getCreatedBy()))
-              .ifNotExists();
+              .value(CassandraAspect.CREATED_BY_COLUMN, literal(aspect.getCreatedBy()));
+      if (updateIfCreatedOn != null && updateIfCreatedOn == 0) {
```
Comment (mattmatravers) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/cassandra/CassandraAspectDao.java:550]: What happens if `updateIfCreatedOn != null` and `updateIfCreatedOn != 0`? We maybe should be able to return a BadRequest via HTTP if so. We at least need to throw an exception back to the client.

File: metadata-io/src/main/java/com/linkedin/metadata/entity/cassandra/CassandraAspectDao.java [hunk 5]
```diff
@@ -559,8 +562,13 @@ private SimpleStatement generateSaveStatement(EntityAspect aspect, boolean inser
 
       Update u = uwa.whereColumn(CassandraAspect.URN_COLUMN).isEqualTo(literal(aspect.getUrn()))
               .whereColumn(CassandraAspect.ASPECT_COLUMN).isEqualTo(literal(aspect.getAspect()))
-              .whereColumn(CassandraAspect.VERSION_COLUMN).isEqualTo(literal(aspect.getVersion()))
-              .ifExists();
+              .whereColumn(CassandraAspect.VERSION_COLUMN).isEqualTo(literal(aspect.getVersion()));
+
+      if (updateIfCreatedOn != null) {
```
Comment (mattmatravers) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/cassandra/CassandraAspectDao.java:567]: Is it feasible for `updateIfCreatedOn == 0` here? I'm not sure where `insert` gets set so I can't reason through this properly. Maybe @RyanHolstien will have an opinion?

File: metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/util/CondUpdateUtil.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+        Map<String, Long> createdOnMap = new HashMap<>();
+        if (condUpdate != null) {
+            Arrays.stream(condUpdate.split(";"))
+                    .forEach(item -> {
+                        String[] values = item.split("=");
+                        if (values.length == 2) {
+                            try {
+                                createdOnMap.put(values[0], Long.parseLong(values[1]));
+                            } catch (Exception e) {
+                                log.warn("Invalid eTag: " + item);
```
Comment (mattmatravers) [inline metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/util/CondUpdateUtil.java:24]: Not sure if eTag is the right name here any more? Maybe use `condition` if you want to be generic?

File: metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/util/CondUpdateUtil.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+@Slf4j
+public class CondUpdateUtil {
+
+    // condUpdate String format: "urn1=createdOn1;urn2=createdOn2;...;urnN=createdOnN"
```
Comment (mattmatravers) [inline metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/util/CondUpdateUtil.java:12]: This implies that each condition is per URN, but I it should be per Aspect i.e. `URN+AspectName1=createdOn1;URN+AspectName2=createdOn2;...;URN+AspectNameN=createdOnN`

File: metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/util/CondUpdateUtil.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+    public static Map<String, Long> extractCondUpdate(String condUpdate) {
+        Map<String, Long> createdOnMap = new HashMap<>();
+        if (condUpdate != null) {
+            Arrays.stream(condUpdate.split(";"))
+                    .forEach(item -> {
+                        String[] values = item.split("=");
+                        if (values.length == 2) {
+                            try {
+                                createdOnMap.put(values[0], Long.parseLong(values[1]));
```
Comment (mattmatravers) [inline metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/util/CondUpdateUtil.java:22]: The official format of [If-Unmodified-Since](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Unmodified-Since) is really awful, but we should either support it or specifically call out the fact that we are deviating from the standard.

```
If-Unmodified-Since: <day-name>, <day> <month> <year> <hour>:<minute>:<second> GMT
```

We should summarise the change in this PR to include the scenarios we are catering for, and also link to the [Conditional Requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests#avoiding_the_lost_update_problem_with_optimistic_locking) document that is the inspiration for our design.

File: metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/util/CondUpdateUtil.java (hunk 7 above)
Comment (mattmatravers) [reply metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/util/CondUpdateUtil.java:12]: @RyanHolstien / @jjoyce0510, is there a standard way to uniquely identify a single Aspect that we can use here?

File: metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/entities/EntitiesController.java [hunk 9]
```diff
@@ -103,15 +107,17 @@ public ResponseEntity<UrnResponseMap> getEntities(
...
     log.info("INGEST PROPOSAL proposal: {}", aspectRequests);
 
     Authentication authentication = AuthenticationContext.getAuthentication();
     String actorUrnStr = authentication.getActor().toUrnStr();
 
+    Map<String, Long> createdOnMap = CondUpdateUtil.extractCondUpdate(condUpdate);
     List<Pair<String, Boolean>> responses = aspectRequests.stream()
         .map(MappingUtil::mapToProposal)
-        .map(proposal -> MappingUtil.ingestProposal(proposal, actorUrnStr,  _entityService, _objectMapper))
+        .map(proposal -> MappingUtil.ingestProposal(proposal, actorUrnStr,  _entityService, _objectMapper, createdOnMap.get(actorUrnStr)))
```
Comment (mattmatravers) [inline metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/entities/EntitiesController.java:120]: Per https://developer.mozilla.org/en-US/docs/Web/HTTP/Conditional_requests#avoiding_the_lost_update_problem_with_optimistic_locking, we will definitely need to return a [Precondition Failed](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/412) status so clients know to retry or adapt their request to the new state.

Are there any places we can do an integration test to show the behaviour from the client side?

**Review by mattmatravers — COMMENTED** [review]
Looking good Djordje, and much easier to reason through and to provide feedback!! There are a few things to work on plus some requests for comment from @RyanHolstien and @jjoyce0510.

File: metadata-io/src/main/java/com/linkedin/metadata/entity/cassandra/CassandraAspectDao.java (hunk 4 above)
Comment (djordje-mijatovic) [reply metadata-io/src/main/java/com/linkedin/metadata/entity/cassandra/CassandraAspectDao.java:550]: Show we throw an exception from CassadraDao class?

Comment (mattmatravers) [reply metadata-io/src/main/java/com/linkedin/metadata/entity/cassandra/CassandraAspectDao.java:550]: We could, although it's possible to spot this sooner in the call stack if it makes more sense. I think we need to catch it somewhere so it's preferable before we get to Cassandra/Ebean specific implementations.

Comment (djordje-mijatovic) [reply metadata-io/src/main/java/com/linkedin/metadata/entity/cassandra/CassandraAspectDao.java:550]: I don't think it is possible as Cassadra implementation doesn't do the database read at all so we can't know until we execute the update query and get the response.

File: metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/entities/EntitiesController.java [hunk 10]
```diff
@@ -103,21 +108,28 @@ public ResponseEntity<UrnResponseMap> getEntities(
 
   @PostMapping(value = "/", produces = MediaType.APPLICATION_JSON_VALUE)
   public ResponseEntity<List<String>> postEntities(
-      @RequestBody @Nonnull List<UpsertAspectRequest> aspectRequests) {
+      @RequestBody @Nonnull List<UpsertAspectRequest> aspectRequests,
```
Comment (jjoyce0510) [inline metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/entities/EntitiesController.java:111]: Can we instead add this as a header to this endpoint? https://github.com/datahub-project/datahub/blob/master/metadata-service/restli-servlet-impl/src/main/java/com/linkedin/metadata/resources/entity/AspectResource.java#L140

This is the ingestProposal

File: metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/entities/EntitiesController.java [hunk 11]
```diff
@@ -103,21 +108,28 @@ public ResponseEntity<UrnResponseMap> getEntities(
...
+              .map(proposal -> MappingUtil.ingestProposal(proposal, actorUrnStr, _entityService, _objectMapper,
+                      createdOnMap.get(actorUrnStr + "+" + proposal.getAspectName())))
+              .collect(Collectors.toList());
+      if (responses.stream().anyMatch(Pair::getSecond)) {
+        return ResponseEntity.status(HttpStatus.CREATED)
+                .body(responses.stream().filter(Pair::getSecond).map(Pair::getFirst).collect(Collectors.toList()));
+      } else {
+        return ResponseEntity.ok(Collections.emptyList());
+      }
+    } catch (PreconditionFailedException e) {
```
Comment (RyanHolstien) [inline metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/entities/EntitiesController.java:130]: This can be pushed into the GlobalControllerExceptionHandler class to avoid wrapping both methods

File: metadata-io/src/main/java/com/linkedin/metadata/entity/AspectDao.java [hunk 12]
```diff
@@ -53,9 +53,10 @@ void saveAspect(
         @Nonnull final Timestamp timestamp,
         @Nonnull final String systemMetadata,
         final long version,
-        final boolean insert);
+        final boolean insert,
+        @Nullable final Long updateIfCreatedOn);
 
-    void saveAspect(@Nonnull final EntityAspect aspect, final boolean insert);
+    void saveAspect(@Nonnull final EntityAspect aspect, final boolean insert, final Long updateIfCreatedOn);
```
Comment (jjoyce0510) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/AspectDao.java:59]: As we discussed -> Let's add an object instead of `updateIfCreatedOn` so we can make the future extension story simpler :)

File: metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java [hunk 13]
```diff
@@ -1884,13 +1892,17 @@ private UpdateAspectResult ingestAspectToLocalDBNoTransaction(
...
+      @Nullable final ExtraIngestParams extraIngestParams) {
 
-    // 2. Compare the latest existing and new.
+    // 1. Compare the latest existing and new.
     final RecordTemplate oldValue =
         latest == null ? null : EntityUtils.toAspectRecord(urn, aspectName, latest.getMetadata(), getEntityRegistry());
     final RecordTemplate newValue = updateLambda.apply(Optional.ofNullable(oldValue));
 
+    // 2. Check if the condition update is met
+    checkConditionalUpdate(latest, extraIngestParams);
```
Comment (mattmatravers) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java:1904]: You could probably push this check further up (above the `final RecordTemplate newValue` line) as we can fail faster that way.

File: metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java [hunk 14]
```diff
@@ -2031,4 +2043,21 @@ private DataPlatformInfo getDataPlatformInfo(Urn urn) {
...
+    if (_aspectDao.supportTransactions() && updateIfCreatedOn != null) {
+      long updateIfCreatedOnValue = updateIfCreatedOn.longValue();
+      if (oldAspect != null && updateIfCreatedOnValue == 0) {
+        throw new PreconditionFailedException("Conditional Update value must be greater then 0 for update operation");
+      }
+      if (oldAspect == null && updateIfCreatedOnValue != 0) {
+        throw new PreconditionFailedException("Conditional Update value must be 0 for insert operation");
+      }
+      if (updateIfCreatedOnValue != 0 && oldAspect != null && oldAspect.getCreatedOn().getTime() != updateIfCreatedOnValue) {
+        throw new PreconditionFailedException("Condition for Conditional Update is not met");
```
Comment (mattmatravers) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java:2059]: Example here: What URN+aspect are we updating? Why couldn't we update it? What is the expected `createdOn` value? What was the actual `createdOn` value? (Who was it changed by?)

File: metadata-io/src/main/java/com/linkedin/metadata/params/ExtraIngestParamsHelper.java [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+@Slf4j
+public class ExtraIngestParamsHelper {
+
+    // condUpdate String format: "urn1+AspectName1=createdOn1;urn2+AspectName2=createdOn2;...;urnN+AspectNameN=createdOnN"
+
+    public static Map<String, Long> extractCondUpdate(String condUpdate) {
```
Comment (mattmatravers) [inline metadata-io/src/main/java/com/linkedin/metadata/params/ExtraIngestParamsHelper.java:14]: This probably needs a JavaDoc to explain the format of the condUpdate parameter. We've explained the `urn1+AspectName1` part but not the format of `createdOn1`, for example.

I'd advise to make an explicit reference to https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-Unmodified-Since and explain that we are NOT following this standard (because it's very hard to follow) but we could follow it in future if some HTTP clients stick religiously to the standard.

File: metadata-io/src/main/java/com/linkedin/metadata/entity/cassandra/CassandraAspectDao.java (hunk 5 above)
Comment (RyanHolstien) [reply metadata-io/src/main/java/com/linkedin/metadata/entity/cassandra/CassandraAspectDao.java:579]: If it's 0 it would go into the above block, not this one. Insert gets set to true when no previous version exists.

File: metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/util/CondUpdateUtil.java (hunk 7 above)
Comment (RyanHolstien) [reply metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/util/CondUpdateUtil.java]: You mean you want all of the Urn,AspectName pairs to have the same AspectName per full identifier? Or can you clarify? If we need a particular aspect that is linked to an Urn you would use the entity key aspect, but it was my understanding this syntax could be used to reference any aspect so it wouldn't need to be consistent. If you need to validate the aspect name is a valid aspect you can validate this using the EntityRegistry

Comment (mattmatravers) [reply metadata-service/openapi-servlet/src/main/java/io/datahubproject/openapi/util/CondUpdateUtil.java]: I just want a consistent way to refer to a unit of change. In DataHub this is an Aspect (attached to a particular entity URN). Is there already a standard way to refer to an aspect such that you could map from a string identifier to an Aspect?

File: metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java [hunk 16]
```diff
@@ -2031,4 +2044,25 @@ private DataPlatformInfo getDataPlatformInfo(Urn urn) {
     }
     return null;
   }
+
+  private void checkConditionalUpdate(EntityAspect oldAspect, ExtraIngestParams extraIngestParams) {
```
Comment (mattmatravers) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java:2048]: For these error messages, it's better to assume the client knows what they are doing and has set the right preconditions, so we're validating the system state, not the contents of the request headers.

For these error messages to be useful, I'm thinking of it from the perspective of an operator that has just sent 5 conditional requests and one or two of them fail. So for each error they will need to know:

1. What URN+aspect was being updated?
2. What was the expected state?
3. What was the actual state?

Given this info, the client can decide how to proceed.

Comment (mattmatravers) [reply metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java:2048]: I'll put an example below but we'd have to add parameters `urn` and `aspectName` to this method to cover the necessary details. We should also add the same detail wherever we `throw new PreconditionFailedException`.

File: metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java [hunk 17]
```diff
@@ -2031,4 +2044,25 @@ private DataPlatformInfo getDataPlatformInfo(Urn urn) {
...
+    Long updateIfCreatedOn = extraIngestParams != null && oldAspect != null
+            ? extraIngestParams.getCreatedOn(oldAspect.getUrn(), oldAspect.getAspect()) : null;
+    if (_aspectDao.supportTransactions() && updateIfCreatedOn != null) {
+      long updateIfCreatedOnValue = updateIfCreatedOn.longValue();
+      if (oldAspect != null && updateIfCreatedOnValue == 0) {
+        throw new PreconditionFailedException("Conditional Update value must be greater then 0 for update operation,"
+                + "expected value is " + oldAspect.getCreatedOn().getTime());
+      }
```
Comment (mattmatravers) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java:2056]: Switching `updateIfCreatedOnValue` to be first as this is our expected state. `oldAspect` is the actual state. I think it flows more easily to think "we expect no existing value, but we found an aspect"
```suggestion
      if (updateIfCreatedOnValue == 0 && oldAspect != null) {
        throw new PreconditionFailedException("Failed to insert " + urn + "+" + aspectName
                + ". Expected no existing aspect, but found one created at " + oldAspect.getCreatedOn().getTime()
                + " by " + oldAspect.getCreatedBy());
      }
```

File: metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java [hunk 18]
```diff
@@ -2031,4 +2044,25 @@ private DataPlatformInfo getDataPlatformInfo(Urn urn) {
...
+      if (oldAspect == null && updateIfCreatedOnValue != 0) {
+        throw new PreconditionFailedException("Conditional Update value must be 0 for insert operation or "
+                + "null if you don't want to apply the condition");
+      }
```
Comment (mattmatravers) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java:2060]:
```suggestion
      if (updateIfCreatedOnValue != 0 && oldAspect == null) {
        throw new PreconditionFailedException("Failed to update " + urn + "+" + aspectName
                + ". Expected an existing aspect created at " + updateIfCreatedOnValue + ", but found none.");
      }
```

File: metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java [hunk 19]
```diff
@@ -2031,4 +2044,25 @@ private DataPlatformInfo getDataPlatformInfo(Urn urn) {
...
+      if (updateIfCreatedOnValue != 0 && oldAspect != null && oldAspect.getCreatedOn().getTime() != updateIfCreatedOnValue) {
+        throw new PreconditionFailedException("Condition for Conditional Update is not met. Received value is "
+                + updateIfCreatedOnValue + ", but expected value was " + oldAspect.getCreatedOn().getTime()
+                + ". Last update was done by " + oldAspect.getCreatedBy());
```
Comment (mattmatravers) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/EntityService.java:2064]:
```suggestion
        throw new PreconditionFailedException("Failed to update " + urn + "+" + aspectName
                + ". Expected an existing aspect created at " + updateIfCreatedOnValue
                + ", but found one created at " + oldAspect.getCreatedOn().getTime()
                + " by " + oldAspect.getCreatedBy());
```

File: metadata-io/src/main/java/com/linkedin/metadata/entity/cassandra/CassandraAspectDao.java [hunk 20]
```diff
@@ -148,10 +150,13 @@ private Map<String, Long> getMaxVersions(@Nonnull final String urn, @Nonnull fin
...
   @Override
-  public void saveAspect(@Nonnull EntityAspect aspect, final boolean insert) {
+  public void saveAspect(@Nonnull EntityAspect aspect, final boolean insert, @Nullable ExtraIngestParams extraIngestParams) {
     validateConnection();
-    SimpleStatement statement = generateSaveStatement(aspect, insert);
-    _cqlSession.execute(statement);
+    SimpleStatement statement = generateSaveStatement(aspect, insert, extraIngestParams);
+    ResultSet resultSet = _cqlSession.execute(statement);
+    if (!resultSet.wasApplied()) {
+      throw new PreconditionFailedException("Condition for Conditional Update is not met");
```
Comment (mattmatravers) [inline metadata-io/src/main/java/com/linkedin/metadata/entity/cassandra/CassandraAspectDao.java:158]: Can this be changed in line with the suggestions above to add urn+aspect details and some expectations to the error message?

**Review by mattmatravers — CHANGES_REQUESTED** [review]
Made some suggestions for error message detail @djordje-mijatovic. I've also sent you my proof-read version of the readme for inclusion, then I'm happy to merge this.
